client/lib/plugins/sanitize-section-content.js

```diff
@@ -0,0 +1,189 @@
+	);
+};
+
+/**
+ * Sanitizes input HTML for security and styling
+ *
+ * @param {String} content unverified HTML
+ * @returns {string} sanitized HTML
+ */
+export const sanitizeSectionContent = content => {
+	const parser = new DOMParser();
+	const doc = parser.parseFromString( content, 'text/html' );
+
+	// this will let us visit every single DOM node programmatically
+	const walker = doc.createTreeWalker(
+		doc.body,
+		NodeFilter.SHOW_ALL,
+		{ acceptNode: () => NodeFilter.FILTER_ACCEPT },
+		false
+	);
+
+	// we don't want to remove nodes while walking the tree
+	// or we'll invite data-race bugs. instead, we'll track
+	// which ones we want to remove then drop them at the end
+	const removeList = [];
+
+	// walk over every DOM node
+	while ( walker.nextNode() ) {
+		const node = walker.currentNode;
+		const tagName = node.nodeName.toLowerCase();
+
+		if ( ! isAllowedTag( tagName ) && ! isValidYoutubeEmbed( node ) ) {
+			removeList.push( node );
+			continue;
+		}
+
+		// strip out anything not explicitly allowed
+		// in the attributes. we want to eliminate
+		// potential cross-site scripting attacks _and_
+		// prevent custom styles from interfering with
+		// our page's own rendering
+
+		const attrRemoveList = [];
+
+		// these aren't Arrays, so iteration is odd
+		for ( let i = 0; node.attributes && i < node.attributes.length; i++ ) {
+			const { name: attrName, value } = node.attributes[ i ];
+
+			// get rid of non-whitelisted attributes
+			if ( ! isAllowedAttr( tagName, attrName ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// for some, specifically links, we don't want to
+			// "sanitize" the values because we could mess them up
+			// for example, encoding a real URL will break it
+			if (
+				( ( 'src' === attrName || 'href' === attrName ) && validUrl.isWebUri( value ) ) ||
+				( 'iframe' === tagName && 'type' === attrName && isValidYoutubeEmbed( node ) )
+			) {
+				continue;
+			}
+
+			// just don't even allow non-url links
+			if ( ( 'href' === attrName || 'src' === attrName ) && ! validUrl.isWebUri( value ) ) {
+				attrRemoveList.push( attrName );
+				continue;
+			}
+
+			// every other other allowed attribute must be sanitized
+			// this is a broad and clumsy stroke but it should be
+			// effective, leaning towards safety instead of
+			// compatability with the remote HTML
+			node.setAttribute( attrName, encodeURIComponent( value ) );
+		}
+
+		// actually remove the attributes
+		attrRemoveList.forEach( name => node.removeAttribute( name ) );
+
+		// of course, all links need to be normalized since
+		// they now exist inside of the Calypso context
+		if ( 'a' === tagName && node.getAttribute( 'href' ) ) {
+			node.setAttribute( 'target', '_blank' );
+			node.setAttribute( 'rel', 'external noopener noreferrer' );
+		}
+	}
+
+	// once done walking the DOM tree
+	// remove the unwanted nodes
+	for ( let i = 0; i < removeList.length; i++ ) {
```

Comment:
can this one can be a forEach since its an Array?

Reply:
abstracted in 82641bed8